### PR TITLE
Added and exception in case of wrong or empty email Also renamed chan…

### PR DESCRIPTION
**Pull request by fanyui**
Added and exception in case of wrong or empty email Also renamed changeUserPasswordUsingActivationKey to changePasswordUsingActivationKey

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 1]
```diff
@@ -723,8 +724,12 @@ public String getSecretQuestion(User user) throws APIException {
 	 */
 	@Override
 	@Transactional(readOnly = true)
-	public User getUserByEmail(String email) throws APIException {	
-		return dao.getUserByEmail(email);
+	public User getUserByEmail(String email) {
+		if (EmailValidator.getInstance().isValid(email)) {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: You're only supposed to check if the email is not null or blank, it's not this method's job to validate the email.

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 2]
```diff
@@ -723,8 +724,12 @@ public String getSecretQuestion(User user) throws APIException {
...
+			return dao.getUserByEmail(email);
+		}
+		throw new APIException("error.email.invalid", (Object[]) null);
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: Please don't pass in null arguments

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 3]
```diff
@@ -1368,6 +1368,13 @@ public void getUserByEmail_shouldGetUserByUsingEmail() {
 		assertNotNull("User with email hank.williams@gmail not found in database", user);
 	}
 	
+	@Test
+	public void getUserByEmail_shouldNotGetUserIfEmailIsNotValid() {
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java]: Should say should fail and not should not get User

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 4]
```diff
@@ -1368,6 +1368,13 @@ public void getUserByEmail_shouldGetUserByUsingEmail() {
...
+		expectedException.expect(APIException.class);
+		expectedException.expectMessage(messages.getMessage("error.email.invalid"));
+		userService.getUserByEmail("");
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java]: You also need tests for null and white space characters

**Comment by wluyima** [comment]
You are supposed to remove some throws clause(s) from some method(s), did you take notes on our call?

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java (hunk 2 above)
Comment (fanyui) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: The constructor which interpretes the string passed from the messages expects two arguments. else the error.email.invalid will just be treated as a normal string

Comment (wluyima) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: Please remove it, take a closer look

Comment (wluyima) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: So you're saying this constructor is the only one that translates messages?

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -723,8 +723,12 @@ public String getSecretQuestion(User user) throws APIException {
 	 */
 	@Override
 	@Transactional(readOnly = true)
-	public User getUserByEmail(String email) throws APIException {	
-		return dao.getUserByEmail(email);
+	public User getUserByEmail(String email) {
+		if (StringUtils.isNotBlank(email)) {
+			return dao.getUserByEmail(email);
+		}
+		throw new APIException("error.email.invalid", (Object[]) null);
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: The error message should say email cannot be null or blank

Comment (wluyima) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: It means you should renamed the message key too to be more intuitive

File: api/src/main/resources/messages.properties [hunk 6]
```diff
@@ -90,6 +90,7 @@ error.context.null=contextDAO is null
 error.rank.notPositiveInteger=Rank can only be a positive integer
 error.email.invalid =Invalid Email address
 error.activationkey.invalid =Invalid user activation Key
+error.email.notNullOrBlank =email cannot be null or blank
```
Comment (wluyima) [inline api/src/main/resources/messages.properties]: Email needs to start with uppercase letter

**Pull request merged**